tmio_read_tag: support skipping known protocol & validity checks

ssailer opened the pull request:
Refactor reading the protocol string from stream and checking it's validity has been moved to an internal helper: `tmio_read_protocol`.

`tmio_read_tag` now checks for allowed tags and skips valid protocol tags.

Some debugging print outs have been added. 

Please check if `tmio_read_tag` sets correct statuses :)

ssailer committed:
tmio_read_tag allows protocol_tag + matching protocol string.
Refactor protocol read into internal helper function.
Add a test for concatenated files. Remove protocol string copy on tmio_open.

fwerner commented on src/tmio.c:
```diff
@@ -408,6 +408,50 @@ TMIO_EHANDSHAKE Error while writing protocol
+            protocol_tag, protocol);
+
+  // Copy hidden messages in protocol string ;)
+  strncpy(stream->protocol, protocol, TMIO_PROTOCOL_SIZE);
```
This breaks the original idea that we only match the protocol prefix, e.g. if we request "FlashCamV1" we should be able to read all "FlashCamV1.x" files.

ssailer replied:
Isn't this the same as in the previous version?
Line 438 here compares until protocol_len of `stream->protocol` and then copies the full string? (see line 495 in the old code). I just removed the ternary operator because the sanitisation of the input already requires that `strlen(stream->protocol) < TMIO_PROTOCOL_SIZE`?
I thought I only added a snarky comment, I could change that though :)
Maybe I miss something.

fwerner commented on src/tmio.c:
```diff
@@ -852,8 +867,23 @@ TMIO_ETIMEDOUT A timeout occured while skipping a data frame
+    if (frame_header < 0) {
+      // have a tag
+      if (frame_header == TMIO_PROTOCOL_TAG) {
+        if (tmio_read_protocol(stream))
```
Doesn't this result in a double read? 🤔

fwerner commented on src/tmio.c:
```diff
@@ -852,8 +867,23 @@ TMIO_ETIMEDOUT A timeout occured while skipping a data frame
+      // have a tag
+      if (frame_header == TMIO_PROTOCOL_TAG) {
+        if (tmio_read_protocol(stream))
+          return -1; // protocol does not match
```
Need to add the status codes set by tmio_read_protocol to the docstring.

fwerner reviewed (commented):
Had a quick look and left some questions. Will come to your office. :)

ssailer commented:
New commit addresses your comments:
- `tmio_read_protocol` does not read a tag anymore, only the string. Usage in `open` and `read_tag` was adapated.
- `strncpy` of the `protocol` was removed.
- Added a crude test, but should be working.

I am still unsure how to treat the statistics of the transient protocol tag: `bytesread` or `bytesskipped`
In the current implementation it's not communicated to the user, that an interleaved `TMIO_PROTOCOL_TAG` arrived. All other statistics are unchanged.
To compare with disksize / streamsize both `read` and `skipped` have to be added anyway.
`skipped` implies additional datafields in the stream not requested.

fwerner committed:
Add access to the latest protocol identifier received from the stream.
Test protocol matching logic and update of stream_protocol for concatenated streams.

fwerner commented:
I've added a getter for the latest protocol identifier received from the stream.

The tests fail, probably because of the statistics tracking that we talked about - need to come to a conclusion on this.

ssailer commented:
Great additions!

The tests failed because the header wasn't updated with the new memory layout of the `tmio_stream` struct and the `writtenbytes` field didn't point to where it was actually stored. (Tests were failing after `tmio_create` which isn't touched in this PR.)

fwerner commented:
I've extended your test to include the protocol matching/update logic.

To me the PR looks ok to merge now, what do you think?

ssailer commented:
I cleaned the test slightly. Looks good to me! Thanks for the work. :)
